self-api/src/main/java/com/selfxdsd/api/CachedResource.java

```diff
@@ -55,65 +56,60 @@
     String etag();
 
     /**
-     * Creation date of the conditional resource.
+     * Creation date of the cached resource.
      *
      * @return LocalDateTime.
      */
     LocalDateTime creationDate();
 
     /**
-     * Factory method that creates a ConditionalResource from a regular
+     * Factory method that creates a CachedResource from a regular
      * Resource.
-     * If the regular can't be conditional
+     * If the regular can't be cached
      * (no ETag header present in its response headers) it returns null.
      *
      * @param uri URI associated with the resource.
      * @param resource Original Resource.
-     * @return ConditionalResource or null.
+     * @return CachedResource or null.
      */
-    static ConditionalResource fromResource(
+    static CachedResource fromResource(
         final URI uri,
         final Resource resource
     ) {
-        final ConditionalResource conditional;
-        if (resource instanceof ConditionalResource) {
-            conditional = (ConditionalResource) resource;
+        final CachedResource cached;
+        if (resource instanceof CachedResource) {
+            final CachedResource candidate = (CachedResource) resource;
+            if (!candidate.uri().equals(uri)) {
+                cached = new FromResource(candidate, candidate.etag(), uri);
+            } else {
+                cached = candidate;
+            }
```

Comment:
@amihaiemil Here a small bug is fixed. Now it makes sure to have the same uri too in order to return the  same instance,
 otherwise it mutates the resource with the new uri. 
Most likely this corner case we'will never encounter in prod...  :D